superbench/benchmarks/micro_benchmarks/ib_loopback_performance.py

```diff
@@ -0,0 +1,214 @@
+            return None
+
+    def __get_arguments_from_env(self):
+        """Read environment variables from runner used for parallel and fill in ib_index and numa_node_index.
+
+        Get 'PROC_RANK'(rank of current process) 'IB_DEVICES' 'NUMA_NODES' environment variables
+        Get ib_index and numa_node_index according to 'NUMA_NODES'['PROC_RANK'] and 'IB_DEVICES'['PROC_RANK']
+        """
+        if os.getenv('PROC_RANK'):
+            rank = int(os.getenv('PROC_RANK'))
+            if os.getenv('IB_DEVICES'):
+                self._args.ib_index = int(os.getenv('IB_DEVICES').split(',')[rank])
+            if os.getenv('NUMA_NODES'):
+                self._args.numa = int(os.getenv('NUMA_NODES').split(',')[rank])
+
+    def _preprocess(self):
+        """Preprocess/preparation operations before the benchmarking.
+
+        Return:
+            True if _preprocess() succeed.
+        """
+        if not super()._preprocess():
+            return False
+
+        self.__get_arguments_from_env()
+
+        # Format the arguments
+        self._args.commands = [command.lower() for command in self._args.commands]
+
+        # Check whether arguments are valid
+        command_mode = ''
+        if self._args.size is None:
+            command_mode = ' -a'
+        else:
+            command_mode = ' -s ' + str(self._args.size)
+
+        for ib_command in self._args.commands:
+            if ib_command not in self.__support_ib_commands:
+                self._result.set_return_code(ReturnCode.INVALID_ARGUMENT)
+                logger.error(
+                    'Unsupported ib command - benchmark: {}, command: {}, expected: {}.'.format(
+                        self._name, ib_command, ' '.join(list(self.__support_ib_commands.keys()))
+                    )
+                )
+                return False
+            else:
+                try:
+                    command = os.path.join(self._args.bin_dir, self._bin_name)
+                    numa_cores = self.__get_numa_cores(self._args.numa)
+                    if len(numa_cores < 4):
+                        self._result.set_return_code(ReturnCode.MICROBENCHMARK_DEVICE_GETTING_FAILURE)
+                        logger.error('Getting numa core devices failure - benchmark: {}.'.format(self._name))
+                        return False
+                    server_core = int(numa_cores[-1])
+                    client_core = int(numa_cores[-3])
+                    command += ' ' + str(server_core) + ' ' + str(client_core)
+                    command += ' ' + self.__support_ib_commands[ib_command]
+                    command += command_mode + ' -F'
+                    command += ' --iters=' + str(self._args.iters)
+                    command += ' -d ' + network.get_ib_devices()[self._args.ib_index]
+                    command += ' -p ' + str(network.get_free_port())
+                    command += ' -x ' + str(self._args.gid_index)
+                    self._commands.append(command)
```

Comment:
add tests for `self._commands`? seems it's not checked currently

Reply:
It can not generate commands successfully, because network.get_ib_devices()[] will raise exception

Reply:
you can mock the return in unit test, e.g.,
```py
from unittest import mock
@mock.patch('superbench.utils.network.get_ib_devices')
def test_ib_loopback(self, mock_ib_devices):
    mock_ib_devices.return_value = ['...']
```

Reply:
updated, thanks